atecc608a_se.c

```diff
@@ -211,6 +211,53 @@ static psa_status_t atecc608a_import_public_key(psa_key_slot_number_t key_slot,
     return status;
 }
 
+static psa_status_t atecc608a_generate_key(psa_key_slot_number_t key_slot,
+                                           psa_key_type_t type,
+                                           psa_key_usage_t usage,
+                                           size_t bits,
+                                           const void *extra,
+                                           size_t extra_size,
+                                           uint8_t *p_pubkey_out,
+                                           size_t pubkey_out_size,
+                                           size_t *p_pubkey_length)
+{
+    const uint16_t key_id = key_slot;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* The hardware has slots 0-15 */
+    if (key_slot > 15)
```

Comment:
I assumed that psa_key_slot_number_t passed to this function is already translated to the hardware slot number, as I believe that that was the plan a couple of weeks ago when I looked into the key slot code update. If so, then 0 is a perfectly fine number for our hardware, as are numbers up to 15, inclusive. There's no reason from this hardware perspective for it to be signed, but as for the driver - I don't have an opinion. I can imagine that there are cryptochips that don't have slot numbers, but an arbitrary identifier of an opaque sort, and then we'd need to have the number translation here, to maintain a consistent API (or just have a void* parameter).
As for the assert vs error - I assumed that there may be cases where the user requests an access to slots incrementally up until an error is returned, and generally, returning an error lets the user to handle the situation in any manner he wants.

Reply:
I don't think we can assume translation from user-facing slot numbers to hardware slot numbers has already occurred at this point. Akin to how the pubkey is in user-facing format on output, the inputs should also be in user-facing format. The driver should control how it does key format translation as well as slot number translation.

Reply:
But this code is not called directly, it will be called through the crypto SE driver API (or rather, PSA will call it), so I assumed that this number will be hidden from the user. 
If what you write would be true, how would slot allocation work? I thought that the user registers or allocates a slot (passing the requested hardware number, since he knows which slot exactly he wants to use) and receives an additional, user-friendly ID. Then, when requesting an operation, PSA translates it before each call to the driver itself.
If that's not the case, each operation would need additional number translation and separate management of slot mapping, and that seems to be a bit of an overkill.

Reply:
The whole concept of slot numbers is still under discussion (https://github.com/ARMmbed/psa-crypto/issues/110). If they exist at all, it'll be some unsigned type, but where 0 is an ordinary value, because typical SEs number their slots from 0. Note that the slot number is not related to the key id (unless it is, but in that case the key id would have to be in a certain small range with a translation `key_id = slot_numer + OFFSET`).

> I would say it makes more sense for the driver itself to perform the translation from user-facing slot numbers to hardware slot numbers

Yes, but… see https://github.com/ARMmbed/psa-crypto/issues/110. The driver lacks storage to remember which slots are occupied, and it's difficult to express the logic for deciding which slot numbers are suitable for which keys.

> I can imagine that there are cryptochips that don't have slot numbers

PKCS#11 doesn't. You can set a label for the key, but you might as well use the key id, there's no need to invent yet another number. At least assuming that there isn't a stringent (&lt;8B) size limitation on the label.